### ISSUE TITLE
Can't retransmit handshake packet 

### ISSUE THREAD

**Issue by scw00** [ngtcp2/ngtcp2#70]
```
I00000000 0x1df83a845332e6e5 frm 2762535630 tx Initial(0x7f) STREAM(0x12) id=0x0 fin=0 offset=0 len=290 uni=0
I00000000 0x1df83a845332e6e5 frm 2762535630 tx Initial(0x7f) PADDING(0x00) len=925
I00000000 0x1df83a845332e6e5 rcv loss_detection_alarm=1522506647784881920 last_hs_tx_pkt_ts=1522506647584881920 alarm_duration=200
I00000002 0x1df83a845332e6e5 con recv packet len=1241
I00000002 0x092d89e403f81fce frm 3904113170 rx Handshake(0x7d) ACK(0x0e) largest_ack=2762535630 ack_delay=0(0) ack_block_count=0
I00000002 0x092d89e403f81fce frm 3904113170 rx Handshake(0x7d) ACK(0x0e) block=[2762535630..2762535630] block_count=0
I00000002 0x092d89e403f81fce rcv latest_rtt=2 min_rtt=2 smoothed_rtt=2.336 rttvar=1.168 max_ack_delay=0
I00000002 0x092d89e403f81fce rcv loss detection alarm canceled
I00000002 0x092d89e403f81fce frm 3904113170 rx Handshake(0x7d) STREAM(0x12) id=0x0 fin=0 offset=0 len=1192 uni=0
I00000002 0x092d89e403f81fce frm 2762535631 tx Handshake(0x7d) ACK(0x0e) largest_ack=3904113170 ack_delay=0(0) ack_block_count=0
I00000002 0x092d89e403f81fce frm 2762535631 tx Handshake(0x7d) ACK(0x0e) block=[3904113170..3904113170] block_count=0
I00000002 0x092d89e403f81fce con recv packet len=799
I00000002 0x092d89e403f81fce frm 3904113171 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00000002 0x092d89e403f81fce frm 2762535632 tx Handshake(0x7d) ACK(0x0e) largest_ack=3904113171 ack_delay=0(0) ack_block_count=0
I00000002 0x092d89e403f81fce frm 2762535632 tx Handshake(0x7d) ACK(0x0e) block=[3904113171..3904113170] block_count=1
I00000002 0x092d89e403f81fce frm 2762535632 tx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=290 len=58 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
QUIC handshake has completed
I00000002 0x092d89e403f81fce frm 2762535633 tx S01(0x1f) ACK(0x0e) largest_ack=3904113171 ack_delay=0(0) ack_block_count=0
I00000002 0x092d89e403f81fce frm 2762535633 tx S01(0x1f) ACK(0x0e) block=[3904113171..3904113170] block_count=1
I00000002 0x092d89e403f81fce frm 2762535633 tx S01(0x1f) STREAM(0x12) id=0x4 fin=0 offset=0 len=1199 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535634 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=1199 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535635 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=2410 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535636 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=3621 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535637 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=4832 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535638 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=6043 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535639 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=7254 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535640 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=8465 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535641 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=9676 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535642 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=10887 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535643 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=12098 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535644 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=13309 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000013 0x092d89e403f81fce rcv loss detection alarm fired
I00000013 0x092d89e403f81fce rcv retransmit unprotected packet 2762535632 sent_ts=1522506647587217920
I00000013 0x092d89e403f81fce rcv handshake_count=1 tlp_count=0 rto_count=0
I00000014 0x092d89e403f81fce rcv loss detection alarm fired
I00000014 0x092d89e403f81fce frm 2762535645 tx S01(0x1f) PING(0x07) len=0 data=
I00000014 0x092d89e403f81fce rcv handshake_count=1 tlp_count=1 rto_count=0
I00000014 0x092d89e403f81fce rcv loss detection alarm fired
I00000014 0x092d89e403f81fce frm 2762535646 tx S01(0x1f) PING(0x07) len=0 data=
I00000014 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=0
I00000014 0x092d89e403f81fce con recv packet len=799
I00000014 0x092d89e403f81fce frm 3904113172 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00000016 0x092d89e403f81fce frm 2762535647 tx S01(0x1f) ACK(0x0e) largest_ack=3904113172 ack_delay=1(187) ack_block_count=0
I00000016 0x092d89e403f81fce frm 2762535647 tx S01(0x1f) ACK(0x0e) block=[3904113172..3904113170] block_count=2
I00000043 0x092d89e403f81fce con recv packet len=799
I00000043 0x092d89e403f81fce frm 3904113173 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00000044 0x092d89e403f81fce frm 2762535648 tx S01(0x1f) ACK(0x0e) largest_ack=3904113173 ack_delay=1(149) ack_block_count=0
I00000044 0x092d89e403f81fce frm 2762535648 tx S01(0x1f) ACK(0x0e) block=[3904113173..3904113170] block_count=3
I00000100 0x092d89e403f81fce con recv packet len=799
I00000100 0x092d89e403f81fce frm 3904113174 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00000101 0x092d89e403f81fce frm 2762535649 tx S01(0x1f) ACK(0x0e) largest_ack=3904113174 ack_delay=1(165) ack_block_count=0
I00000101 0x092d89e403f81fce frm 2762535649 tx S01(0x1f) ACK(0x0e) block=[3904113174..3904113170] block_count=4
I00000202 0x092d89e403f81fce rcv loss detection alarm fired
I00000202 0x092d89e403f81fce frm 2762535650 tx S01(0x1f) PING(0x07) len=0 data=
I00000202 0x092d89e403f81fce frm 2762535651 tx S01(0x1f) PING(0x07) len=0 data=
I00000202 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=1
I00000212 0x092d89e403f81fce con recv packet len=799
I00000212 0x092d89e403f81fce frm 3904113175 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00000213 0x092d89e403f81fce frm 2762535652 tx S01(0x1f) ACK(0x0e) largest_ack=3904113175 ack_delay=1(167) ack_block_count=0
I00000213 0x092d89e403f81fce frm 2762535652 tx S01(0x1f) ACK(0x0e) block=[3904113175..3904113170] block_count=5
I00000402 0x092d89e403f81fce rcv loss detection alarm fired
I00000402 0x092d89e403f81fce frm 2762535653 tx S01(0x1f) PING(0x07) len=0 data=
I00000402 0x092d89e403f81fce frm 2762535654 tx S01(0x1f) PING(0x07) len=0 data=
I00000402 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=2
I00000436 0x092d89e403f81fce con recv packet len=799
I00000436 0x092d89e403f81fce frm 3904113176 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00000438 0x092d89e403f81fce frm 2762535655 tx S01(0x1f) ACK(0x0e) largest_ack=3904113176 ack_delay=1(181) ack_block_count=0
I00000438 0x092d89e403f81fce frm 2762535655 tx S01(0x1f) ACK(0x0e) block=[3904113176..3904113170] block_count=6
I00000802 0x092d89e403f81fce rcv loss detection alarm fired
I00000802 0x092d89e403f81fce frm 2762535656 tx S01(0x1f) PING(0x07) len=0 data=
I00000802 0x092d89e403f81fce frm 2762535657 tx S01(0x1f) PING(0x07) len=0 data=
I00000802 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=3
I00000884 0x092d89e403f81fce con recv packet len=799
I00000884 0x092d89e403f81fce frm 3904113177 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00000886 0x092d89e403f81fce frm 2762535658 tx S01(0x1f) ACK(0x0e) largest_ack=3904113177 ack_delay=1(168) ack_block_count=0
I00000886 0x092d89e403f81fce frm 2762535658 tx S01(0x1f) ACK(0x0e) block=[3904113177..3904113170] block_count=7
I00001602 0x092d89e403f81fce rcv loss detection alarm fired
I00001602 0x092d89e403f81fce frm 2762535659 tx S01(0x1f) PING(0x07) len=0 data=
I00001602 0x092d89e403f81fce frm 2762535660 tx S01(0x1f) PING(0x07) len=0 data=
I00001602 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=4
I00001779 0x092d89e403f81fce con recv packet len=799
I00001779 0x092d89e403f81fce frm 3904113178 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00001781 0x092d89e403f81fce frm 2762535661 tx S01(0x1f) ACK(0x0e) largest_ack=3904113178 ack_delay=1(200) ack_block_count=0
I00001781 0x092d89e403f81fce frm 2762535661 tx S01(0x1f) ACK(0x0e) block=[3904113178..3904113170] block_count=8
I00003203 0x092d89e403f81fce rcv loss detection alarm fired
I00003203 0x092d89e403f81fce frm 2762535662 tx S01(0x1f) PING(0x07) len=0 data=
I00003203 0x092d89e403f81fce frm 2762535663 tx S01(0x1f) PING(0x07) len=0 data=
I00003203 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=5
I00003568 0x092d89e403f81fce con recv packet len=799
I00003568 0x092d89e403f81fce frm 3904113179 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00003569 0x092d89e403f81fce frm 2762535664 tx S01(0x1f) ACK(0x0e) largest_ack=3904113179 ack_delay=1(173) ack_block_count=0
I00003569 0x092d89e403f81fce frm 2762535664 tx S01(0x1f) ACK(0x0e) block=[3904113179..3904113170] block_count=9
I00006406 0x092d89e403f81fce rcv loss detection alarm fired
I00006406 0x092d89e403f81fce frm 2762535665 tx S01(0x1f) PING(0x07) len=0 data=
I00006406 0x092d89e403f81fce frm 2762535666 tx S01(0x1f) PING(0x07) len=0 data=
I00006406 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=6
I00007145 0x092d89e403f81fce con recv packet len=799
I00007145 0x092d89e403f81fce frm 3904113180 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00007147 0x092d89e403f81fce frm 2762535667 tx S01(0x1f) ACK(0x0e) largest_ack=3904113180 ack_delay=1(171) ack_block_count=0
I00007147 0x092d89e403f81fce frm 2762535667 tx S01(0x1f) ACK(0x0e) block=[3904113180..3904113170] block_count=10
I00012803 0x092d89e403f81fce rcv loss detection alarm fired
I00012803 0x092d89e403f81fce frm 2762535668 tx S01(0x1f) PING(0x07) len=0 data=
I00012803 0x092d89e403f81fce frm 2762535669 tx S01(0x1f) PING(0x07) len=0 data=
I00012803 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=7
I00014301 0x092d89e403f81fce con recv packet len=799
I00014301 0x092d89e403f81fce frm 3904113181 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00014302 0x092d89e403f81fce frm 2762535670 tx S01(0x1f) ACK(0x0e) largest_ack=3904113181 ack_delay=1(168) ack_block_count=0
I00014302 0x092d89e403f81fce frm 2762535670 tx S01(0x1f) ACK(0x0e) block=[3904113181..3904113170] block_count=11
I00025613 0x092d89e403f81fce rcv loss detection alarm fired
I00025613 0x092d89e403f81fce frm 2762535671 tx S01(0x1f) PING(0x07) len=0 data=
I00025613 0x092d89e403f81fce frm 2762535672 tx S01(0x1f) PING(0x07) len=0 data=
I00025613 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=8
I00028609 0x092d89e403f81fce con recv packet len=799
I00028609 0x092d89e403f81fce frm 3904113182 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00028611 0x092d89e403f81fce frm 2762535673 tx S01(0x1f) ACK(0x0e) largest_ack=3904113182 ack_delay=1(169) ack_block_count=0
I00028611 0x092d89e403f81fce frm 2762535673 tx S01(0x1f) ACK(0x0e) block=[3904113182..3904113170] block_count=12
I00051225 0x092d89e403f81fce rcv loss detection alarm fired
I00051225 0x092d89e403f81fce frm 2762535674 tx S01(0x1f) PING(0x07) len=0 data=
I00051225 0x092d89e403f81fce frm 2762535675 tx S01(0x1f) PING(0x07) len=0 data=
I00051225 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=9
I00057226 0x092d89e403f81fce con recv packet len=799
I00057226 0x092d89e403f81fce frm 3904113183 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00057228 0x092d89e403f81fce frm 2762535676 tx S01(0x1f) ACK(0x0e) largest_ack=3904113183 ack_delay=1(171) ack_block_count=0
I00057228 0x092d89e403f81fce frm 2762535676 tx S01(0x1f) ACK(0x0e) block=[3904113183..3904113170] block_count=13
Timeout
I00087257 0x092d89e403f81fce frm 2762535677 tx S01(0x1f) CONNECTION_CLOSE(0x02) error_code=NO_ERROR(0) reason_len=0


```

**Comment by scw00** [comment]
https://github.com/ngtcp2/ngtcp2/blob/master/examples/server.cc#L1358-L1384

we should call another write_pkt if it is the handshake packet.

**Comment by scw00** [comment]
I think write the retransmit pkt may be better.
https://github.com/ngtcp2/ngtcp2/blob/master/lib/ngtcp2_conn.c#L4891-L4919

**Comment by tatsuhiro-t** [comment]
ngtcp2_conn_handshake does retransmission.

**Comment by scw00** [comment]
Hmm, I don't think we can use ngtcp2_conn_handshake since this function will callback to `send_client_handshake` and run out of the chandshake_ array.

And after first call of ngtcp2_conn_handshake we only call this function when server send something here .

https://github.com/ngtcp2/ngtcp2/blob/d2fb61b701b2c5977dec29e36573c2c67249e90d/examples/server.cc#L1357-L1394

It is out of retransmit block.

**Comment by scw00** [comment]
I think it is the root cause of the log in first comment.

**Comment by scw00** [comment]
The log in first comment is client side, since I can't reproduce this issue with logging.

**Comment by tatsuhiro-t** [comment]
Are you using latest master branch?
do_handshake will be called unless handshake has not finished yet.  There is no return statement unless there is an error.

**Comment by scw00** [comment]
>Are you using latest master branch?
Yes it is latest master .

And I updated log.

Let me explain what happens with updated log .

First, we send the initial packet . and recv the handshake packet (3904113170), and send ack packet(2762535631). 
```
I00000002 0x092d89e403f81fce frm 3904113170 rx Handshake(0x7d) ACK(0x0e) largest_ack=2762535630 ack_delay=0(0) ack_block_count=0
I00000002 0x092d89e403f81fce frm 3904113170 rx Handshake(0x7d) ACK(0x0e) block=[2762535630..2762535630] block_count=0
I00000002 0x092d89e403f81fce rcv latest_rtt=2 min_rtt=2 smoothed_rtt=2.336 rttvar=1.168 max_ack_delay=0
I00000002 0x092d89e403f81fce rcv loss detection alarm canceled
I00000002 0x092d89e403f81fce frm 3904113170 rx Handshake(0x7d) STREAM(0x12) id=0x0 fin=0 offset=0 len=1192 uni=0
I00000002 0x092d89e403f81fce frm 2762535631 tx Handshake(0x7d) ACK(0x0e) largest_ack=3904113170 ack_delay=0(0) ack_block_count=0
I00000002 0x092d89e403f81fce frm 2762535631 tx Handshake(0x7d) ACK(0x0e) block=[3904113170..3904113170] block_count=0
I00000002 0x092d89e403f81fce con recv packet len=799
```

Then server sends another handshake packet(3904113171).
```
I00000002 0x092d89e403f81fce frm 3904113171 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
```

**And now we can export the FINISHED message(2762535633) and send it to server. But this packet was lost**
```
I00000002 0x092d89e403f81fce frm 2762535633 tx S01(0x1f) ACK(0x0e) largest_ack=3904113171 ack_delay=0(0) ack_block_count=0
I00000002 0x092d89e403f81fce frm 2762535633 tx S01(0x1f) ACK(0x0e) block=[3904113171..3904113170] block_count=1
I00000002 0x092d89e403f81fce frm 2762535633 tx S01(0x1f) STREAM(0x12) id=0x4 fin=0 offset=0 len=1199 uni=0
```

we send the protected data and wait for the firing of alarm
```
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535634 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=1199 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535635 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=2410 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535636 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=3621 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535637 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=4832 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535638 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=6043 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535639 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=7254 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535640 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=8465 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535641 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=9676 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535642 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=10887 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535643 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=12098 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
I00000002 0x092d89e403f81fce frm 2762535644 tx S01(0x1f) STREAM(0x16) id=0x4 fin=0 offset=13309 len=1211 uni=0
I00000002 0x092d89e403f81fce rcv loss_detection_alarm=1522506647597217920 last_hs_tx_pkt_ts=1522506647587217920 alarm_duration=10
```

Now we detect the last client handshake message was lost and try to retransmit it, but it seems something has pended the retransmitting of handshake.
```
I00000013 0x092d89e403f81fce rcv loss detection alarm fired
I00000013 0x092d89e403f81fce rcv retransmit unprotected packet 2762535632 sent_ts=1522506647587217920
I00000013 0x092d89e403f81fce rcv handshake_count=1 tlp_count=0 rto_count=0
I00000014 0x092d89e403f81fce rcv loss detection alarm fired
I00000014 0x092d89e403f81fce frm 2762535645 tx S01(0x1f) PING(0x07) len=0 data=
I00000014 0x092d89e403f81fce rcv handshake_count=1 tlp_count=1 rto_count=0
I00000014 0x092d89e403f81fce rcv loss detection alarm fired
I00000014 0x092d89e403f81fce frm 2762535646 tx S01(0x1f) PING(0x07) len=0 data=
I00000014 0x092d89e403f81fce rcv handshake_count=1 tlp_count=2 rto_count=0
I00000014 0x092d89e403f81fce con recv packet len=799
I00000014 0x092d89e403f81fce frm 3904113172 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0
I00000016 0x092d89e403f81fce frm 2762535647 tx S01(0x1f) ACK(0x0e) largest_ack=3904113172 ack_delay=1(187) ack_block_count=0
I00000016 0x092d89e403f81fce frm 2762535647 tx S01(0x1f) ACK(0x0e) block=[3904113172..3904113170] block_count=2
I00000043 0x092d89e403f81fce con recv packet len=799
I00000043 0x092d89e403f81fce frm 3904113173 rx Handshake(0x7d) STREAM(0x16) id=0x0 fin=0 offset=1192 len=760 uni=0

```

**Comment by scw00** [comment]
This comment based on the code .

We call `Client::on_write`  (in run function) when we are running the client program. and call do_handshake to send client initial packet. 

Then client recv the server handshake packet(3904113170 and 3904113171) `Client::on_read `and client exports the  client finished message and sends it to server(2762535633) and lost this packet. `Client::on_write`-> `Client::do_handshake` and schedule an alarm.

We may write some protected data, but I think it dose not a matter.

 Then alarm fired .

we get into the `retransmitcb` https://github.com/ngtcp2/ngtcp2/blob/d2fb61b701b2c5977dec29e36573c2c67249e90d/examples/server.cc#L498-L535

We call the first on_write with retransmit flag when alarm firing.

**Comment by scw00** [comment]
Dig it more, it seems the window limit the retransmit

https://github.com/ngtcp2/ngtcp2/blob/d2fb61b701b2c5977dec29e36573c2c67249e90d/lib/ngtcp2_conn.c#L1904-L1909

**Comment by scw00** [comment]
Why do we use the ping frame rather than write the lost packets ?

https://github.com/ngtcp2/ngtcp2/blob/d2fb61b701b2c5977dec29e36573c2c67249e90d/lib/ngtcp2_conn.c#L4891-L4919

**Comment by tatsuhiro-t** [comment]
Yes, it is due to congestion window depleted.
It looks like we have to ignore congestion window when retransmitting handshake packets.  I'm not sure recovery draft say this explicitly.

We send PING because draft is recommended to do so.  It is better because probe packet is not counted to congestion control window.   Also it is only done in Short (non-handshake) packet, and it does not solve this particular issue.

**Comment by tatsuhiro-t** [comment]
Please try latest master.  This issue should be fixed now.

**Comment by scw00** [comment]
spec says
>A packet sent on an RTO alarm MUST NOT be blocked by the sender's congestion controller.

**Comment by tatsuhiro-t** [comment]
It is RTO, not TLP.  Anyway, sending PING or not is not relevant to this issue.

**Comment by scw00** [comment]
Thanks 

Yep,  I've tested the latest master, it works but I find another issue.

**Comment by scw00** [comment]
`
